pkg/sources/generator/tickgen.go

```diff
@@ -312,12 +312,60 @@ func (mg *memgen) Start() <-chan struct{} {
 	return mg.forwarder.Start()
 }
 
+func (mg *memgen) NewWorker(ctx context.Context, rate int) func(chan time.Time, chan struct{}) {
+
+	return func(tickChan chan time.Time, done chan struct{}) {
+		for {
+			select {
+			case <-ctx.Done():
+				log.Info("Context.Done is called. emptying any pending ticks")
+				for len(tickChan) > 0 {
+					<-tickChan
+				}
+				done <- struct{}{}
+				return
+			case ts := <-tickChan:
+				tickgenSourceCount.With(map[string]string{metrics.LabelVertex: mg.name, metrics.LabelPipeline: mg.pipelineName})
+				// we would generate all the keys in a round robin fashion
+				// even if there are multiple pods, all the pods will generate same keys in the same order.
+				// alternatively, we could also think about generating a subset of keys per pod.
+				t := ts.UnixNano()
+				for i := 0; i < rate; i++ {
+					for k := int32(0); k < mg.keyCount; k++ {
+						key := fmt.Sprintf("key-%d", k)
+						payload := mg.genfn(mg.msgSize, mg.value, t)
+						r := record{data: payload, offset: time.Now().UTC().UnixNano(), key: key}
+						select {
+						case <-ctx.Done():
+							log.Info("Context.Done is called. returning from the inner function")
+							return
+						case mg.srcchan <- r:
+						}
+					}
+				}
+			}
+		}
+	}
+}
+
 // generator fires once per time unit and generates records and writes them to the channel
 func (mg *memgen) generator(ctx context.Context, rate int, timeunit time.Duration) {
 	go func() {
-		var rCount int32 = 0
-		// we are capping the limit at 10000 msgs / second
-		var limit = int32(10000 / rate)
+		// capping the rate to 10000 msgs/sec
+		if rate > 10000 {
+			rate = 10000
+		}
+
+		tickChan := make(chan time.Time, math.MaxInt32)
+		doneChan := make(chan struct{})
+		childCtx, cancel := context.WithCancel(ctx)
```

Comment:
```suggestion
		childCtx, cancel := context.WithCancel(ctx)
		defer cancel()
```
you can remove the explicit cancel in https://github.com/numaproj/numaflow/pull/661/files#diff-1145efea11df28ccc255ebaef330c426040ffeeaef5a9e6a7978713e837ec837R377

Reply:
wanted to wait until the other routine exits  cleanly.